gobblin-core/src/main/java/gobblin/writer/AvroHdfsTimePartitionedWriter.java

```diff
@@ -0,0 +1,284 @@
+
+    // Initialize the partitionLevel
+    this.partitionLevel =
+        this.properties.getProp(getWriterPartitionLevel(this.branch), ConfigurationKeys.DEFAULT_WRITER_PARTITION_LEVEL);
+
+    // Initialize the timestampToPathFormatter
+    this.timestampToPathFormatter =
+        DateTimeFormat.forPattern(
+            this.properties.getProp(getWriterPartitionPattern(this.branch),
+                ConfigurationKeys.DEFAULT_WRITER_PARTITION_PATTERN)).withZone(
+            DateTimeZone.forID(this.properties.getProp(ConfigurationKeys.WRITER_PARTITION_TIMEZONE,
+                ConfigurationKeys.DEFAULT_WRITER_PARTITION_TIMEZONE)));
+
+    // Check that ConfigurationKeys.WRITER_PARTITION_COLUMN_NAME has been specified and is properly formed
+
+    Preconditions.checkArgument(this.properties.contains(getWriterPartitionColumnName(this.branch)),
+        "Missing required property " + ConfigurationKeys.WRITER_PARTITION_COLUMN_NAME);
+
+    this.partitionColumnName = this.properties.getProp(getWriterPartitionColumnName(this.branch));
+    Optional<Schema> writerPartitionColumnSchema = AvroUtils.getFieldSchema(this.schema, this.partitionColumnName);
+
+    Preconditions.checkArgument(writerPartitionColumnSchema.isPresent(), "The column " + this.partitionColumnName
+        + " specified by " + ConfigurationKeys.WRITER_PARTITION_COLUMN_NAME + " is not in the writer input schema");
+
+    Preconditions.checkArgument(writerPartitionColumnSchema.get().getType().equals(Schema.Type.LONG), "The column "
+        + this.partitionColumnName + " specified by " + ConfigurationKeys.WRITER_PARTITION_COLUMN_NAME
+        + " must be of type " + Schema.Type.LONG);
+  }
+
+  @Override
+  public void write(GenericRecord record) throws IOException {
+
+    // Retrieve the value of the field specified by this.partitionColumnName
+    Optional<Object> writerPartitionColumnValue = AvroUtils.getFieldValue(record, this.partitionColumnName);
+    Preconditions.checkState(writerPartitionColumnValue.isPresent());
+
+    Path writerOutputPath = getPathForColumnValue((Long) writerPartitionColumnValue.get());
+
+    // If the path is not in pathToWriterMap, construct a new DataWriter, add it to the map, and write the record
+    // If the path is in pathToWriterMap simply retrieve the writer, and write the record
+    if (!this.pathToWriterMap.containsKey(writerOutputPath)) {
+
+      LOG.info("Creating a new DataWriter for path: " + new Path(this.baseFilePath, writerOutputPath));
+
+      DataWriter<GenericRecord> avroHdfsDataWriter = createAvroHdfsDataWriterForPath(writerOutputPath);
+
+      this.pathToWriterMap.put(writerOutputPath, avroHdfsDataWriter);
+      avroHdfsDataWriter.write(record);
+    } else {
+      this.pathToWriterMap.get(writerOutputPath).write(record);
+    }
+  }
+
+  @Override
+  public void commit() throws IOException {
+    for (DataWriter<GenericRecord> dataWriter : this.pathToWriterMap.values()) {
+      dataWriter.commit();
```

Comment:
Should catch `IOException` so every writers get a chance to commit even if some fails and rethrow the `IOException` if any writer fails.